src/linear_algebra/general_linear_group.lean

```diff
@@ -0,0 +1,200 @@
+ * `matrix.GL_plus.GL_pos` gives the subgroup of matrices with
+ positive determinant (over a linear ordered ring)
+
+## Implementation notes
+
+
+## References
+
+
+## Tags
+
+matrix group, group, matrix inverse
+-/
+
+namespace matrix
+universes u v
+open_locale matrix
+open linear_map
+
+section
+
+variables (n : Type u) [decidable_eq n] [fintype n] (R : Type v) [comm_ring R]
+
+/-- `GL n R` is the group of `n` by `n` `R`-matrices with unit determinant.
+Defined as a subtype of matrices-/
+abbreviation general_linear_group (R : Type*) [comm_ring R] : Type* := units (matrix n n R)
+
+notation `GL` := general_linear_group
+
+lemma GL.is_unit_det  (A : GL n R) : is_unit ((A : matrix n n R).det) :=
+begin
+  rw ←matrix.is_unit_iff_is_unit_det,
+  exact units.is_unit _
+end
+
+end
+namespace general_linear_group
+
+variables {n : Type u} [decidable_eq n] [fintype n] {R : Type v} [comm_ring R]
+
+/--The `GL n R` and `general_linear_group R n` groups are multiplicatively equivalent-/
+def to_lin : (GL n R) ≃* (linear_map.general_linear_group R (n → R)) :=
+units.map_equiv to_lin_alg_equiv'.to_mul_equiv
+
+/--Given a matrix with invertible determinant we get an element of `GL n R`-/
+ def mk' (A : matrix n n R) (h : invertible (det A)) : GL n R :=
+unit_of_det_invertible A
+
+/--Given a matrix with unit determinant we get an element of `GL n R`-/
+noncomputable def mk'' (A : matrix n n R) (h : is_unit (det A)) : GL n R :=
+nonsing_inv_unit A h
+
+instance coe_fun : has_coe_to_fun (GL n R) :=
+{ F   := λ _, n → n → R,
+  coe := λ A, A.val }
+
+lemma ext_iff (A B : GL n R) : A = B ↔ (∀ i j, A i j = B i j) :=
+begin
+  dsimp at *,
+  fsplit,
+  work_on_goal 0 { intros a i j, induction a, refl },
+  intros a,
+  ext,
+  solve_by_elim,
+end
+
+@[ext] lemma ext (A B : GL n R) : (∀ i j, A i j = B i j) → A = B :=
+(ext_iff A B).mpr
+
+/-- Applying `matrix.nonsing_inv` to the coercion is the same as coercing the group inverse. -/
+lemma nonsing_inv_coe_eq_coe_inv (A : GL n R) : (↑A : matrix n n R)⁻¹ = ↑(A⁻¹) :=
+begin
+  letI := A.invertible,
+  exact (inv_eq_nonsing_inv_of_invertible ↑A).symm,
+end
+
+section coe_lemmas
+
+variables (A B : GL n R)
+
+@[simp] lemma coe_fn_eq_coe : ⇑A = (↑A : matrix n n R) := rfl
+
+@[simp] lemma mul_val : ↑(A * B) = A ⬝ B := rfl
+
+@[simp] lemma one_val : ↑(1 : GL n R) = (1 : matrix n n R) := rfl
+
+@[simp] lemma inv_val : ↑(A⁻¹) = nonsing_inv A :=
+begin
+  have:=nonsing_inv_coe_eq_coe_inv A, unfold_coes at *, rw ← this, refl,
+end
+
+@[simp] lemma to_lin'_mul : to_lin' (A * B) = (to_lin' A).comp (to_lin' B) :=
+matrix.to_lin'_mul A B
+
+@[simp] lemma to_lin'_one : to_lin' (1 : GL n R).1 = linear_map.id :=
+matrix.to_lin'_one
+
+end coe_lemmas
+
+instance SL_to_GL : has_coe (special_linear_group n R) (GL n R) :=
+⟨λ A, ⟨↑A, ↑(A⁻¹), congr_arg coe (mul_right_inv A), congr_arg coe (mul_left_inv A)⟩⟩
+
+@[simp] lemma GL_vals (A : GL n R) : ∀ i j, A i j = A.1 i j :=
+begin
+unfold_coes,
+intros i j,
+refl,
+end
```

Comment:
I think now that you have `coe_fn_eq_coe` you shouldn't need this lemma

Reply:
Good point. I've removed it now.